### DISCUSSION TITLE
Power Query inside of TE3?

### DISCUSSION THREAD

**Discussion by InsightABI** [TabularEditor/TabularEditor3#498] [Ideas]
I was saying to my co-worker today:

"What I was trying to say today is that I would love to right-mouse-click a shared expression or an M partition property in TE3 and choose _Edit In Power Query_. Then be able to build in PQ (like we do in Desktop) and then have the resulting M query be saved back to the model."

Is this being considered as an enhancement? 

@otykier

**Comment by otykier** [comment]
While we would love to provide such functionality, we don’t have the bandwidth to recreate the full PQ experience in TE3, and it’s not possible to use Microsoft’s Power Query editor for legal reasons. The best we can do (which is already on our roadmap), is to provide a better editor for M expressions (syntax checking, IntelliSense, etc.)

**Reply by InsightABI** [reply]
I go to the Microsoft Partner Community call every week (I think you have demo'd in it before).   Every Quarter the session is a Ask Me Anything with Amir Netz.  Next time we have it, I'll ask him if there is any plan to make Power Query open-source or if Microsoft has considered licensing it to Kapacity -- for free, of course ; ) -- so that it can work seamlessly inside of TE3.  Can't hurt to ask!  

@paulturley  Respectfully, this Idea was raised for PQ in TE3.  Your question on bulk renaming might be more appropriately posted in the [Q&A](https://github.com/TabularEditor/TabularEditor3/discussions/categories/q-a) section as a new discussion.  : )  BTW, I've read many of your articles and really enjoy them.

**Reply by fausttiger007** [reply]
Do we open the BIM file in Visual Studio Pro, make our power query edits or imports,  then reopen the BIM in  TE3 ?

Tony

**Reply by otykier** [reply]
> Do we open the BIM file in Visual Studio Pro, make our power query edits or imports, then reopen the BIM in TE3 ?
> 
> Tony

If you need the full PQ editor, then yes. You could also simply create a new model in VS or Power BI Desktop, or even Excel, open the Power Query editor and create your data transformations. When done, copy the M expression over to your model in Tabular Editor.

**Comment by paulturley** [comment]
One of the features I have used frequently in TE2 is the ability to bulk rename measures. How is this facilitated in TE3?

**Reply by otykier** [reply]
Hi Paul. Multi-select the objects you wish to rename then hit F2. A batch rename dialog should then pop up.

**Reply by paulturley** [reply]
Got it. It looks like you must have added this since my last update on the machine I'm using. I see it on the right-click menu as well.
Thanks!
